client/vaultclient/vaultclient.go

```diff
@@ -370,6 +370,11 @@ func (c *vaultClient) RenewToken(token string, increment int) (<-chan error, err
 		increment: increment,
 	}
 
+	// Add renewal request to the heap to start tracking the token.
+	c.lock.Lock()
+	c.heap.Push(renewalReq, time.Now())
+	c.lock.Unlock()
```

Comment:
Should we remove the original [`heap.Push`](https://github.com/hashicorp/nomad/pull/18985/files#diff-61b0f81ca4505d23bb2adf7025d2c36c4e02232a2f13ac39dce779617c954416R518-R523) or at least change this comment?

```go
// If the identifier is not already tracked, this is a first
// renewal request. In this case, add an entry into the heap
// with the next renewal time.
```

Actually, is that whole branch now an unreachable state because the [`if c.isTracked(req.id)`](https://github.com/hashicorp/nomad/pull/18985/files#diff-61b0f81ca4505d23bb2adf7025d2c36c4e02232a2f13ac39dce779617c954416R479) conditional will always be true?

Reply:
Hum...good point. Looking at the code again, I think this change can actually add a _new_ racing condition where the token is renewed twice!

When `RenewToken()` pushes the token to heap the `run()` loop may pick it up for renewal right away, but then `RenewToken()` calls `c.renew()` for an immediate renewal, resulting in a double renew. 

I don't _think_ that's a problem but it's certainly unnecessary. 

I think we could make it so only the `run()` loop actually renew tokens and `RenewToken()` blocks until the renewal request completes, but that could add some latency to task start.

Another option is to revert the new `!c.isTracked` check and just adjust the error so it's considered `fatal` and we only get one error max. 

The `select` race condition that it tries to address seems rather unlikely to happen, so a single sporadic error may not be too bad? 

Another option would be to have different behaviours if `c.renew()` is called from `RenewToken()` or `run()`, like a new `renewUntracked()` function that doesn't check the heap 🤔

Reply:
> I think we could make it so only the run() loop actually renew tokens and RenewToken() blocks until the renewal request completes, but that could add some latency to task start.

I like this idea, because trying to call `renew` from two different places seems to be the root of the race. The only potential added latency would come from another token/lease being renewed at the same time as the task starts, so that it hits the `run` loop while that loop is otherwise occupied. But I'm pretty sure we already have that same latency today, because only one token/lease can be renewed at the same time because `renew` takes the mutex.

Reply:
Cool! Implemented this in https://github.com/hashicorp/nomad/pull/18998.

Since it's a fair amount of changes I think it may be better to avoid backporting it. I then updated this PR to only include the extra error check which will prevent the multiple renewals and is much safer to backport.